scripts/parse_live.py

```diff
@@ -144,56 +186,36 @@ def work(pageids):
         if snippets_rows:
             article_row = (pageid, url, title)
             rows.append({'article': article_row, 'snippets': snippets_rows})
-
-    def insert(cursor, r):
-        cursor.execute('''
-            INSERT INTO articles VALUES(%s, %s, %s)''', r['article'])
-        cursor.executemany('''
-            INSERT IGNORE INTO snippets VALUES(%s, %s, %s, %s)''',
-            r['snippets'])
-
-        # We can't allow data to be truncated for HTML snippets, as that can
-        # completely break the UI, so we detect truncation warnings and get rid
-        # of the corresponding data.
-        warnings = cursor.execute('SHOW WARNINGS')
-        truncated_snippets = []
-        for _, _, message in cursor.fetchall():
-            m = DATA_TRUNCATED_WARNING_RE.match(message)
-            if m is None:
-                # Not a truncation, ignore (it's already logged)
-                continue
-            # MySQL warnings index rows starting at 1
-            idx = int(m.groups()[0]) - 1
-            truncated_snippets.append((r['snippets'][idx][0],))
-        if len(truncated_snippets) < len(r['snippets']):
-            cursor.executemany('''
-                DELETE FROM snippets WHERE id = %s''', truncated_snippets)
-        else:
-            # Every single snippet was truncated, remove the article itself
-            cursor.execute('''DELETE FROM articles WHERE page_id = %s''',
-                (r['article'][0],))
     # Open a short-lived connection to try to avoid the limit of 20 per user:
     # https://phabricator.wikimedia.org/T216170
     db = chdb.init_scratch_db()
     for r in rows:
         db.execute_with_retry(insert, r)
 
-def parse_live(pageids, timeout):
+def parse_live(pageids, cd_data, timeout):
     backdir = tempfile.mkdtemp(prefix = 'citationhunt_parse_live_')
     pool = multiprocessing.Pool(
         # The number of processes per CPU is pretty much made up. The processes
         # are I/O bound as most of the time is spent querying the Wikipedia API.
         processes = multiprocessing.cpu_count() * 8,
         initializer = initializer, initargs = (backdir,))
 
-    # Make sure we query the API 32 pageids at a time
+    # Make sure we query the API 32 pageids/revids at a time
     tasks = []
     batch_size = 32
-    pageids_list = list(pageids)
-    for i in range(0, len(pageids), batch_size):
-        tasks.append(pageids_list[i:i+batch_size])
 
-    result = pool.map_async(work, tasks)
+    if pageids:
+        pageids_list = list(pageids)
+        for i in range(0, len(pageids), batch_size):
+            tasks.append(pageids_list[i:i+batch_size])
+        result = pool.map_async(work, tasks)
+    else:
+        revids_list = list(set([revid for revid, sentence in cd_data]))
+        for i in range(0, len(revids_list), batch_size):
+            revids_batch = revids_list[i:i+batch_size]
+            tasks.append([row for row in cd_data if row[0] in revids_batch])
```

Comment:
Hm, maybe I'm misreading, but it sounds like you don't need to create `revids_list`?

```python
for i in range(0, len(cd_data), batch_size):
  tasks.append(cd_data[i:i+batch_size])
```

?

Reply:
Here because one revid may correspond to multiple sentences, some revids in cd_data are repeated (also that's why I use set). In this way, we will process batch_size revids and unfixed number (>=batch_size) of sentences at a time.
And in your version, we will process batch_size sentences at a time, and they may all come from the same revid.

Reply:
Oh, I see, thanks you. This is worth leaving a comment for, then. Something like "A revid might have multiple sentences and appear repeatedly in the rows."

I also worry a little about the volume of data here: if we have hundreds of thousands of rows, those list-to-set conversions and list comprehensions could get quite slow (especially the `tasks.append` line which goes over all of the data for each batch).

We might want to go with something like:

```python
def chunk_cd_data(cd_data, batch_size):
 tasks = []
 size = 0
 batch = []
 for revid, sentence in cd_data:
    # Assume cd_data was sorted by revid.
    if not batch or revid != batch[-1][0]:
      size += 1
    batch.append((revid, sentence))
    if size == batch_size:
      tasks.append(batch)
      size = 0
      batch = []
  return tasks
```

...which only goes over the data once. But ignore me if you're already testing this with real data and it's not turning out to be a problem, the first step of optimizing is always to measure and see if there's a problem, and where, before trying to write "fast" code :)

Reply:
This is a good point. I've tested with real data before, but not big size, only 37488 rows. There is no problem so far. I think I'll need to test with larger data and see what happens.
And I'll leave a comment. Thank you!